### PR TITLE
feat!: Add validation of Autoscaler config against JSON schema

### PR DESCRIPTION
**Pull request by nielm**
Fully Deprecate minNodes, maxNodes from configuration

Validate the config against a JSON schema.

Add command line config validator for JSON and GKE configMaps

BREAKING CHANGE: previous configs may fail due to stricter configuration validation.

### CODE REVIEW

**Comment by nielm** [comment]
Potential fix for #318

**Comment by alexlo03** [comment]
Hello, this looks good
1) can you please document what the behavior would be if the configuration is invalid? looks like "no-op plus error log"
2) maybe there should be a metric outcome in addition to the logs outcome

**Comment by nielm** [comment]
> Hello, this looks good
> 
> 1. can you please document what the behavior would be if the configuration is invalid? looks like "no-op plus error log"
> 2. maybe there should be a metric outcome in addition to the logs outcome

yes, the poller component will detect the invalid configuration and this will result in a no-op along with: 
* [An error log message from the poller](https://github.com/nielm/autoscaler/blob/f14beb984b3b17097f6d7a65ded95b718794e1b8/src/poller/poller-core/index.js#L618)
* incrementing the `poller/requests-failed` metric
* sending an error to Cloud Error Reporting in the cloud console. 

This will happen every $polling_interval until the config is fixed.

File: package.json [hunk 1]
```diff
@@ -28,7 +28,8 @@
     "test-e2e": "pushd terraform/cloud-functions/per-project/test && go test -run . -timeout 60m --tags=e2e && popd",
     "typecheck": "tsc --project jsconfig.json --maxNodeModuleJsDepth 0 --noEmit",
     "unified-job": "node -e \"require('./src/unifiedScaler').main()\"",
-    "update-all": "npm update -S"
+    "update-all": "npm update -S",
+    "validateConfigFile": "node -e \"require('./src/poller/poller-core/config-validator').main()\" -- "
```
Comment (henrybell) [inline package.json:32]: Perhaps make this kebab-case for consistency?

File: src/poller/README.md [hunk 2]
```diff
@@ -69,6 +69,14 @@ configuration parameters are defined in YAML in a [Kubernetes ConfigMap][configm
 See the [configuration section][autoscaler-home-config] in the home page for
 instructions on how to change the payload.
 
+The Autoscaler JSON (for Cloud functions) or YAML (for GKE) configuration can be
+validated by running the command:
+
+```shell
+npm install
+npm run validateConfigFile -- path/to/config_file
```
Comment (henrybell) [inline src/poller/README.md:77]: kebab-case to match above

File: src/poller/poller-core/config-validator.js [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+      'utf-8',
+    );
+
+    const schema = JSON.parse(configSchema);
+    this.ajvConfigValidator = tmpAjv.compile(schema);
+    this.ajv = tmpAjv;
+  }
+
+  /**
+   * Validates the given object against the Memorystore Config schema.
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:77]: Nit: MCA is leaking

File: src/poller/poller-core/config-validator.js [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+          this.ajv.errorsText(this.ajvConfigValidator.errors, {
+            separator: '\n',
+            dataVar: 'SpannerConfig',
+          }),
+      );
+    }
+  }
+
+  /**
+   * Parses the given string as JSON and validate it against the SannerConfig
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:96]: Nit: Spanner

File: src/poller/poller-core/config-validator.js [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    } catch (e) {
+      throw new Error(`Invalid JSON in Autoscaler configuration: ${e}`);
+    }
+    await this.assertValidConfig(configJson);
+    return configJson;
+  }
+}
+
+/**
+ * Validates the specified Memorystore Autoscaler JSON configuration file.
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:115]: Nit: MCA leakage

File: src/poller/poller-core/config-validator.js [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      );
+    } else {
+      console.error(`Processing of config in file ${filename} failed: ${e}`);
+    }
+    throw new Error(`${filename} Failed validation`);
+  }
+}
+
+/**
+ * Validates all the Memorystore Autoscaler YAML config files specified in the
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:138]: Nit: MCA leakage

File: src/poller/poller-core/config-validator.js [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+/**
+ * Validates a configuration file passed in on the command line.
+ */
+function main() {
+  if (
+    process.argv.length <= 1 ||
+    process.argv[1] === '-h' ||
+    process.argv[1] === '--help'
+  ) {
+    console.log('Usage: validateConfigFile CONFIG_FILE_NAME');
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:200]: Kebab-case for consistency?

File: src/poller/poller-core/test/config-validator.test.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
+/* Copyright 2020 Google LLC
```
Comment (henrybell) [inline src/poller/poller-core/test/config-validator.test.js:1]: Nit: 2024

File: src/poller/poller-core/test/resources/good-mutli-config.yaml
Comment (henrybell) [inline src/poller/poller-core/test/resources/good-mutli-config.yaml:1]: Rename `good-mutli-config.yaml` -> `good-multi-config.yaml`

File: terraform/gke/README.md [hunk 9]
```diff
@@ -697,6 +697,13 @@ following the instructions above.
     cat ${AUTOSCALER_ROOT}/autoscaler-config/autoscaler-config.yaml
     ```
 
+3.  Validate the contents of the YAML configuraration file:
+
+    ```sh
+    npm install
+    npm run validateConfigFile -- ${AUTOSCALER_ROOT}/autoscaler-config/autoscaler-config.yaml
```
Comment (henrybell) [inline terraform/gke/README.md:704]: Kebab-case as above

File: src/poller/poller-core/test/resources/good-mutli-config.yaml
Comment (henrybell) [reply src/poller/poller-core/test/resources/good-mutli-config.yaml:1]: Also should we have a `good-multi-config.json`?

File: src/poller/poller-core/config-validator.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+  if (configMap.kind !== 'ConfigMap') {
+    console.error(`${filename} is not a GKE ConfigMap`);
+    throw new Error(`${filename} is not a GKE ConfigMap`);
+  }
+
+  let success = true;
+  for (const configMapFile of Object.keys(configMap.data)) {
+    const configMapData = configMap.data[configMapFile];
+    try {
+      const memorystoreConfig = yaml.load(configMapData);
```
Comment (henrybell) [inline src/poller/poller-core/config-validator.js:165]: Nit: MCA leakage here and just below

**Review by henrybell — APPROVED** [review]

**Pull request merged**